python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,716 @@
+
+    def _task_execute_delete(self, *args, **kwargs):
+        """
+        Method that gets executed in a background task/thread to perform a Shotgun
+        delete call
+
+        :param *args:       Unnamed arguments to be passed to the delete() call
+        :param **kwargs:    Named arguments to be passed to the delete() call 
+        :returns:           Dictionary containing the 'action' together with result
+                            returned by the delete() call
+        """
+        sg_res = self._bundle.shotgun.delete(*args, **kwargs)
+        return {"action":"delete", "sg_result":sg_res}
+
+    def _task_execute_method(self, method, method_args, method_kwargs):
+        """
+        Method that gets executed in a background task/thread to execute a method
+        with a thread-specific shotgun connection.
+
+        :param method:          The method to be run asyncronously
+        :param method_args:     Arguments to be passed to the method
+        :param method_kwargs:   Named arguments to be passed to the method 
+        :returns:               Dictionary containing the 'action' together with the result
+                                returned by the method
+        """
+        res = method(self._bundle.shotgun, *method_args, **method_kwargs)
+        return {"action":"method", "result":res}
+
+    def _task_check_thumbnail(self, url, load_image):
+        """
+        Check to see if a thumbnail exists for the specified url.  If it does then it is returned.
+
+        :param url:         The url to return the cached path for
+        :param load_image:  If True then if the thumbnail is found in the cache then the file will 
+                            be loaded into a QImage
+        :returns:           A dictionary containing the cached path for the specified url and a QImage
+                            if load_image is True and the thumbnail exists in the cache.
+        """
+        # first look up the path in the cache:
+        thumb_path = ShotgunDataRetriever._get_thumbnail_path(url, self._bundle)
+        thumb_image = None
+        if thumb_path and os.path.exists(thumb_path):
+            if load_image:
+                # load the thumbnail into a QImage:
+                thumb_image = QtGui.QImage()
+                thumb_image.load(thumb_path)
+        else:
+            thumb_path = None
+
+        return {"action":"check_thumbnail", "thumb_path":thumb_path, "image":thumb_image}
+
+    def _task_download_thumbnail(self, thumb_path, url, entity_type, entity_id, field, load_image, **kwargs):
+        """
+        Download the thumbnail for the specified entity type, id and field.  This downloads the
+        thumbnail into the thumbnail cache directory and returns the cached path.
+
+        If thumb_path already contains a path then this method does nothing and just returns the path 
+        without further checking/work.
+
+        :param thumb_path:  Path to an existing thumbnail or None.
```

Comment:
sorry this is most likely just my brain being very jetlagged, but where is `thumb_path` coming from? it's not contained in the dictionary passed to `add_task`